### PR TITLE
CTCP-4146: 

### PR DESCRIPTION
**Pull request by joecefnhedog**

### CODE REVIEW

File: app/pages/locationOfGoods/AddContactYesNoPage.scala [hunk 1]
```diff
@@ -34,10 +34,12 @@ case object AddContactYesNoPage extends QuestionPage[Boolean] {
   override def route(userAnswers: UserAnswers, departureId: String, mode: Mode): Option[Call] =
     Some(routes.AddContactYesNoController.onPageLoad(departureId, mode))
 
-  override def cleanup(value: Option[Boolean], userAnswers: UserAnswers): Try[UserAnswers] =
+  override def cleanup(value: Option[Boolean], userAnswers: UserAnswers): Try[UserAnswers] = {
+    val jsonPathForPhone15 = JsPath \ "Consignment" \ "LocationOfGoods" \ "ContactPerson"
     value match {
-      case Some(false) => userAnswers.remove(LocationOfGoodsContactSection)
+      case Some(false) => userAnswers.remove(LocationOfGoodsContactSection, Set(jsonPathForPhone15))
```
Comment (joecefnhedog) [inline app/pages/locationOfGoods/AddContactYesNoPage.scala:40]: I think we can remove the set here

File: app/pages/locationOfGoods/IdentificationPage.scala [hunk 2]
```diff
@@ -34,11 +34,23 @@ trait BaseIdentificationPage extends QuestionPage[LocationOfGoodsIdentification]
...
+      JsPath \ "Consignment" \ "LocationOfGoods" \ "UNLocode",
+      JsPath \ "Consignment" \ "LocationOfGoods" \ "CustomsOffice",
+      JsPath \ "Consignment" \ "LocationOfGoods" \ "GNSS",
+      JsPath \ "Consignment" \ "LocationOfGoods" \ "EconomicOperator",
+      JsPath \ "Consignment" \ "LocationOfGoods" \ "Address",
+      JsPath \ "Consignment" \ "LocationOfGoods" \ "PostcodeAddress"
+    )
     value match {
-      case Some(_) => userAnswers.remove(QualifierOfIdentificationDetailsSection).flatMap(cleanup)
+      case Some(_) => userAnswers.remove(QualifierOfIdentificationDetailsSection, listOfPaths).flatMap(cleanup)
```
Comment (joecefnhedog) [inline app/pages/locationOfGoods/IdentificationPage.scala:50]: setOfPaths

File: app/utils/AnswersHelper.scala [hunk 3]
```diff
@@ -60,9 +60,9 @@ class AnswersHelper(
     convertToReference: String => Future[T],
     valueToConvert: Option[String]
   )(implicit userAnswers: UserAnswers, rds: Reads[T]): Future[Option[T]] =
-    for {
-      getFrom70         <- Future.successful(userAnswers.get(page))
-      referenceDataCall <- valueToConvert.map(convertToReference).sequence
-    } yield getFrom70.orElse(referenceDataCall)
+    userAnswers.get(page) match {
+      case Some(value) => Future.successful(Some(value))
+      case None        => valueToConvert.map(convertToReference).sequence
```
Comment (joecefnhedog) [inline app/utils/AnswersHelper.scala:65]: we want to be lazy on the referenceData call hence the change

File: app/utils/LocationOfGoodsAnswersHelper.scala [hunk 4]
```diff
@@ -62,37 +68,193 @@ class LocationOfGoodsAnswersHelper(
...
+  def postCodeAddress: Option[SummaryListRow] = getAnswerAndBuildRow[PostalCodeAddress](
+    page = PostalCodePage,
+    formatAnswer = formatAsPostalCode,
+    prefix = "locationOfGoods.postalCode",
+    findValueInDepartureData = message => message.Consignment.LocationOfGoods.flatMap(_.PostcodeAddress.map(_.toPostalCode)),
+    id = Some("change-location-of-goods-postalCode")
+  )
+
   def locationOfGoodsSection: Future[Section] = {
+    implicit val ua: UserAnswers = userAnswers
```
Comment (joecefnhedog) [inline app/utils/LocationOfGoodsAnswersHelper.scala:181]: fetchValue takes an implicit userAnswers, happy to remove this if req

File: app/controllers/locationOfGoods/AddressController.scala [hunk 5]
```diff
@@ -51,41 +50,66 @@ class AddressController @Inject() (
...
-            val preparedForm = request.userAnswers.get(AddressPage) match {
-              case None        => form(isPostalCodeRequired)
-              case Some(value) => form(isPostalCodeRequired).fill(value)
-            }
+        val getCountry: Option[Country] = request.userAnswers
+          .get(CountryPage)
+          .orElse(
+            request.userAnswers.departureData.Consignment.LocationOfGoods.flatMap(
+              _.Address.map(
+                x => Country(CountryCode(x.country), "")
```
Comment (joecefnhedog) [inline app/controllers/locationOfGoods/AddressController.scala:65]: we don't need the whole country, better than making one with an empty description?

File: app/controllers/locationOfGoods/IdentificationController.scala [hunk 6]
```diff
@@ -55,46 +55,70 @@ class IdentificationController @Inject() (
...
+        request.userAnswers
+          .get(InferredLocationTypePage)
+          .map(_.`type`)
+      )
+      .orElse(
+        request.userAnswers.departureData.Consignment.LocationOfGoods.map(
+          x => x.typeOfLocation
+        )
+      )
+    getLocationType
```
Comment (jackheslop96) [inline app/controllers/locationOfGoods/IdentificationController.scala:122]: Do we need to assign this to a val if we're just returning it?

File: app/controllers/locationOfGoods/IdentificationController.scala [hunk 7]
```diff
@@ -55,46 +55,70 @@ class IdentificationController @Inject() (
...
+      .get(LocationTypePage)
+      .map(_.`type`)
+      .orElse(
+        request.userAnswers
+          .get(InferredLocationTypePage)
+          .map(_.`type`)
+      )
+      .orElse(
+        request.userAnswers.departureData.Consignment.LocationOfGoods.map(
+          x => x.typeOfLocation
```
Comment (jackheslop96) [inline app/controllers/locationOfGoods/IdentificationController.scala:119]: Very minor but can use syntactic sugar here _.typeOfLocation

File: app/controllers/locationOfGoods/AddressController.scala [hunk 8]
```diff
@@ -51,41 +50,46 @@ class AddressController @Inject() (
 
   private type Request = SpecificDataRequestProvider1[Country]#SpecificDataRequest[_]
 
-  private def country(implicit request: Request): Country = request.arg
-
-  private def form(isPostalCodeRequired: Boolean)(implicit request: Request): Form[DynamicAddress] =
-    formProvider("locationOfGoods.address", isPostalCodeRequired)
+  private def form(isPostalCodeRequired: Boolean)(implicit request: MandatoryDataRequest[_]): Form[DynamicAddress] =
```
Comment (jackheslop96) [inline app/controllers/locationOfGoods/AddressController.scala:53]: Is the `Request` type still being used?

File: app/controllers/locationOfGoods/IdentificationController.scala (hunk 6 above)
Comment (joecefnhedog) [reply app/controllers/locationOfGoods/IdentificationController.scala:122]: intelij refactor method's idea, removed.

File: app/controllers/locationOfGoods/AddressController.scala (hunk 8 above)
Comment (joecefnhedog) [reply app/controllers/locationOfGoods/AddressController.scala:53]: Nope, removed

**Review by jackheslop96 — APPROVED** [review]

**Pull request merged**